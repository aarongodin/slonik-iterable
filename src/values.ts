import { sql, SqlTokenType } from "slonik"

import { translateValue } from "./translate"

export function fromObject(obj: Record<string, any>, translate = translateValue): SqlTokenType {
  const values = Object.keys(obj).map((identifier) => {
    const value = obj[identifier]
    return translate(identifier, value)
  })

  return sql.join(values, sql`, `)
}
