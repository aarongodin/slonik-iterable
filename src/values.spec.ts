import { sql } from "slonik"

import * as values from "./values"

describe("values helpers", () => {
  describe("fromObject()", () => {
    it("should return the slonik construct for a set of object properties", () => {
      const actual = values.fromObject({
        first: "first value",
        second: 222,
      })

      expect(actual).toMatchSnapshot()
    })

    it("should return the slonik construct for a set of object properties with translation", () => {
      const actual = values.fromObject(
        {
          first: "first value",
          specialType: { isaJson: true },
        },
        (column, value) => {
          if (column === "specialType") {
            return sql.json(value)
          } else {
            return value
          }
        },
      )

      expect(actual).toMatchSnapshot()
    })

    it("should return the slonik construct for a set of object properties with translation that operates on the column name", () => {
      const actual = values.fromObject(
        {
          first: "first value",
          specialType: { isaJson: true },
        },
        (column, value) => {
          if (column === "specialType") {
            return ["special_type", sql.json(value)]
          } else {
            return value
          }
        },
      )

      expect(actual).toMatchSnapshot()
    })
  })
})
