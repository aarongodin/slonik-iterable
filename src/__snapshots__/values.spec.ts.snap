// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`values helpers fromObject() should return the slonik construct for a set of object properties 1`] = `
Object {
  "glue": Object {
    "sql": ", ",
    "type": "SLONIK_TOKEN_SQL",
    "values": Array [],
  },
  "members": Array [
    "first value",
    222,
  ],
  "type": "SLONIK_TOKEN_LIST",
}
`;

exports[`values helpers fromObject() should return the slonik construct for a set of object properties with translation 1`] = `
Object {
  "glue": Object {
    "sql": ", ",
    "type": "SLONIK_TOKEN_SQL",
    "values": Array [],
  },
  "members": Array [
    "first value",
    Object {
      "type": "SLONIK_TOKEN_JSON",
      "value": Object {
        "isaJson": true,
      },
    },
  ],
  "type": "SLONIK_TOKEN_LIST",
}
`;
